File: week11/20/pr3_min_cost_to_connect_all_points.cpp
#include <vector>
using namespace std;
#include <queue>
#include <math.h>

class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        int n = points.size();
        if(n == 1) return 0;

        int used = 0;
        vector<int> visited(n,false);
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> mq;
        mq.push({0,0});
        int ans = 0;
        while(used < n){
            int cost = mq.top().first;
            int v = mq.top().second;
            mq.pop();
            if(visited[v]) continue;
            ans += cost;
            visited[v] = true;
            used++;

            for(int i=0;i<n;i++){
                if(!visited[i]){
                    mq.push({abs(points[i][0]-points[v][0])+abs(points[i][1]-points[v][1]),i});
                }
            }
        }
        return ans;
    }
};

File: week11/20/pr3_min_cost_to_connect_all_points.py
from heapq import heappop, heappush
from typing import List

class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        if n <= 1:
            return 0

        # 맨해튼 거리 계산 함수
        def manhattan(point1, point2):
            return abs(point1[0] - point2[0]) + abs(point1[1] - point2[1])

        ans = 0
        min_heap = [(0, 0)]  # (cost, point index)
        in_mst = [False] * n
        edges_used = 0

        while edges_used < n:
            cost, u = heappop(min_heap)
            if in_mst[u]:
                continue
            
            # 현재 점을 MST에 추가
            in_mst[u] = True
            ans += cost
            edges_used += 1

            # 새로운 점에서 다른 점들까지의 비용을 계산하고 힙에 추가
            for v in range(n):
                if not in_mst[v]:
                    heappush(min_heap, (manhattan(points[u], points[v]), v))

        return ans
